warehouse/internal/loadfiles/loadfiles.go

```diff
@@ -0,0 +1,348 @@
+		messages = nil
+		batchStartIdx := i
+		batchEndIdx := j
+
+		g.Go(func() error {
+			responses := <-ch
+			lf.Logger.Infof("[WH]: Received responses for staging files %d:%d for %s:%s from PgNotifier", toProcessStagingFiles[batchStartIdx].ID, toProcessStagingFiles[batchEndIdx-1].ID, destType, destID)
+
+			var loadFiles []model.LoadFile
+			var successfulStagingFileIDs []int64
+			for _, resp := range responses {
+				// Error handling during generating_load_files step:
+				// 1. any error returned by pgnotifier is set on corresponding staging_file
+				// 2. any error effecting a batch/all the staging files like saving load file records to wh db
+				//    is returned as error to caller of the func to set error on all staging files and the whole generating_load_files step
+				if resp.Status == "aborted" {
+					lf.Logger.Errorf("[WH]: Error in generating load files: %v", resp.Error)
+					sampleError = fmt.Errorf(resp.Error)
+					err = lf.StageRepo.SetErrorStatus(ctx, resp.JobID, sampleError)
+					if err != nil {
+						return fmt.Errorf("set staging file error status: %w", err)
+					}
+					continue
+				}
+				var output []WorkerJobResponse
+				err = json.Unmarshal(resp.Output, &output)
+				if err != nil {
+					return fmt.Errorf("unmarshalling response from pgnotifier: %w", err)
+				}
+				if len(output) == 0 {
+					lf.Logger.Errorf("[WH]: No LoadFiles returned by wh worker")
+					continue
+				}
+				for i := range output {
+					loadFiles = append(loadFiles, model.LoadFile{
+						TableName:             output[i].TableName,
+						Location:              output[i].Location,
+						TotalRows:             output[i].TotalRows,
+						ContentLength:         output[i].ContentLength,
+						StagingFileID:         output[i].StagingFileID,
+						DestinationRevisionID: output[i].DestinationRevisionID,
+						UseRudderStorage:      output[i].UseRudderStorage,
+						SourceID:              job.Upload.SourceID,
+						DestinationID:         job.Upload.DestinationID,
+						DestinationType:       job.Upload.DestinationType,
+					})
+				}
+
+				successfulStagingFileIDs = append(successfulStagingFileIDs, resp.JobID)
+			}
+
+			err = lf.LoadRepo.Insert(ctx, loadFiles)
+			if err != nil {
+				return fmt.Errorf("inserting load files: %w", err)
+			}
+			err = lf.StageRepo.SetStatuses(ctx, successfulStagingFileIDs, warehouseutils.StagingFileSucceededState)
+			if err != nil {
+				return fmt.Errorf("setting staging file status to succeeded: %w", err)
+			}
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return 0, 0, err
+	}
+
+	loadFiles, err := lf.LoadRepo.GetByStagingFiles(ctx, repo.StagingFileIDs(toProcessStagingFiles))
+	if err != nil {
+		return 0, 0, fmt.Errorf("getting load files: %w", err)
+	}
+	if len(loadFiles) == 0 {
+		return 0, 0, fmt.Errorf(`no load files generated. Sample error: %v`, sampleError)
+	}
+
+	if !slices.IsSortedFunc(loadFiles, func(a, b model.LoadFile) bool {
```

Comment:
Didn't get why we this check?

Reply:
Because the order of the files plays a vital role in computing the start/end id, I wanted to add safeguards and not only rely upon the model returning them in the correct order.